### ISSUE TITLE
Questions about dpt_order, branch and figure size.

### ISSUE THREAD

**Issue by ShuhuaGao** [scverse/scanpy#27]
Hi,
I have tried the  [Moignard](https://github.com/theislab/scanpy_usage/tree/master/170501_moignard15) early blood development example and your results can be reproduced. However, I have some questions as follows.

1. What does the dpt_order in the final result mean?
I thought it should be the order of cells according to their pseudotime. However, after checking the "smp.csv" in the generated data, it seems there is no correspondence between _dpt_order_ and _dpt_pseudotime_. For instance, the order of 3 cells with the smallest _dpt_pseudotime_ are 3308, 1813, 3301 respectively (instead of 0, 1, 2).

2. How to adjust the figure size of sc.pl.dpt?
I have tried to put `plt.figure(figsize=(15,15))`, but it seems that it doesn't work. Maybe we have to plot each figure manually. 

3. Can we get information about branch assignment?
In this example, there are two branches. How can we know which branch each cell belongs to if possible? Do we have to find this indirectly with the _dpt_group_ information?

Thanks.

**Comment by ShuhuaGao** [comment]
About dpt_order in the Moignard example, if we plot the pseudotime vs. dpt-order at last manually like follows, 
`plt.figure(figsize=(8,8))
plt.plot(adata.smp['dpt_order'], adata.smp['dpt_pseudotime'])
plt.xlabel('dpt-order')
plt.ylabel('pseudotime')`
we'll get
![download 1](https://user-images.githubusercontent.com/20141984/28011291-687c20e0-6594-11e7-8068-472b4ab4d8a8.png)
which is totally chaotic, unlike the one we get through `sc.pl.dpt`.

Anyway, did I misunderstand the concept of dpt-order or is there a bug about this?

**Comment by falexwolf** [comment]
Sorry for the slow response.

The `'dpt_order'` is an index vector that orders cells according to pseudotime within each group. So, `adata[adata.smp['dpt_order']]` will be ordered in this way. You should not expect correlation of the index vector, which has very weird-looking entries, and `dpt_pseudotime`

I recommend only using `'dpt_pseudotime'` and `'dpt_groups'`. With `adata_0 = adata[adata.smp['dpt_groups']=='0']` you select the data points within group '0', which you could, e.g. order by pseudotime manually using `adata_0 = adata_0[np.argsort(adata_0.smp['dpt_pseudotime'])]`.

To resize figures, use matplolib's `rcParams['savefig.dpi']` or the convenience function that writes that parameter `sc.settings.set_dpi(150)`. You can fully control the figure parameters by changing the rcParams after importing scanpy. Scanpy, similarly to Seaborn, changes a few default settings when imported, see [here](https://github.com/theislab/scanpy/blob/2cea8341e28eb8d0658f62d010631f77465e16d7/scanpy/plotting/utils.py#L55-L95).

Does this answer your questions?

**Comment by ShuhuaGao** [comment]
Many thanks. I got your idea about setting the figure properties now. However, regarding dpt-order, I am still a little confused. By running the Moignard example, we can get the following figure,
![1](https://user-images.githubusercontent.com/20141984/28050880-153abf14-6633-11e7-971f-30b841dba50e.png)

In this figure, is the dpt-order the same one as we have acquired by `sc.write('data/moignard15.csv', adata)`. It seems not because in the generated _smp.csv_, the dpt_order is not varying from 0 to about 700 for group 0,  which is shown in the figure. Therefore, the above figure cannot be produced with the dpt_order in the generated _smp.csv_, right? We have to recompute the order manually by what you have said, like `adata_0 = adata_0[np.argsort(adata_0.smp['dpt_pseudotime'])]`. If so, then I got it. In short, we cannot use the dpt_order field in the generated _smp.csv_ directly.

**Comment by falexwolf** [comment]
```
import matplotlib.pyplot as pl
pl.plot(range(adata.n_smps), adata.smp['dpt_pseudotime'][adata.smp['dpt_order']])
pl.show()
```
This is how you can reproduce the figure and use the `'dpt_order'` field without having to sort the pseudotime yourself again. 😄

**Comment by ShuhuaGao** [comment]
Thanks for your patience. I am almost there. 😄
With your code, I can reproduce the figure on the bottom, i.e., pseudotime-dpt order. However, how about the top one? After we write the result, in the smp.csv file, the data looks like following. (I ordered by dpt_group in Excel.)
![2](https://user-images.githubusercontent.com/20141984/28059812-ff899ea4-6657-11e7-90a7-2857969775ba.png)
Apparently, the cells with dpt_order such as 971, 966 and 960  belong to group 0, which is different from the top sub-figure, where they are assigned into group 1. How shall we explain this?

**Comment by falexwolf** [comment]
Ok, now I see why you are confused! The x-axis in the figure above is somehow confusingly labeled 'dpt_order' even though it simply corresponds to `range(adata.n_smps)`, as evident from the code snippet I posted above; posting it again with the weird axis labeling that confuses you
```
import matplotlib.pyplot as pl
pl.plot(range(adata.n_smps), adata.smp['dpt_pseudotime'][adata.smp['dpt_order']])
pl.xlabel('dpt_order')
pl.show()
```

I definitely have to change this naming. Still it is meaningful that `adata.smp['dpt_order']` is an index vector that "generates" the order. OK, very soon, I could do the following. I'll rename the index array that generates the order `adata.smp['dpt_generate_order']` and the order you see in the plot `adata.smp['dpt_ordering_id']`, if you think this is better. So, one probably fast computation of this is the following
```
ordering_id = np.zeros(adata.n_smps, dtype=int)
for count, idx in adata.smp['dpt_generate_order']:
     ordering_id[idx] = count
```

What do you think?

**Comment by ShuhuaGao** [comment]
Well, I think it is better to be consistent with the paper [Diffusion pseudotime robustly reconstructs lineage branching](https://www.nature.com/nmeth/journal/v13/n10/full/nmeth.3971.html) since this paper first introduces the diffusion pseudotime concept.  In Figure 1 (c) of this paper, there is a _DPT order_.  It seems the dpt order in this paper is just a global rank for each individual cell according to their pseudotime. Therefore, I suggest that the adata.smp['dpt_order'] and the one in the figure should have the same meaning, though IMHO dpt_order only matters for cells on the same branch.  If we extract cells by their dpt_group, then the dpt_order is still applicable even though it is not continuous now. 

In short, I think a dpt_order defined as the global rank by pseudotime like the one in the original paper is more understandable.  By the way, if there are multiple branches in the diffusion map, is there some way to assign the cells to a certain branch? That is to say, can we provide an _adata.smp['branch']_ field?

**Comment by falexwolf** [comment]
The dpt branch field you are look for is the `adata.smp['dpt_groups']` field.

I completely agree that it should be consistent with our paper; that's what it was meant to be. Also the figure above is consistent with that, it's just that the 'dpt_order' field currently is an index field that generates the order, but does not provide ids for it. I'll make this more transparent in the future.

Cheers!

**Comment by ShuhuaGao** [comment]
OK, happy ending. Thanks for all these discussions.
